src/steamship/agents/context/chathistory.py

```diff
@@ -0,0 +1,225 @@
+from __future__ import annotations
+
+import logging
+from typing import Dict, List, Optional, Union
+
+import tiktoken
+
+from steamship import File, MimeTypes, SteamshipError
+from steamship.base.client import Client
+from steamship.data import TagKind
+from steamship.data.block import Block
+from steamship.data.tags import Tag
+from steamship.data.tags.tag_constants import ChatTag, DocTag, RoleTag, TagValueKey
+
+
+class ChatHistory:
+    class Config:
+        arbitrary_types_allowed: True
+
+    """A ChatHistory is a wrapper of a File ideal for ongoing interactions between a user and a virtual assistant."""
+
+    file: File
+
+    def __init__(self, file: File):
+        """This init method is intended only for private use within the class. See `Chat.create()`"""
+        self.file = file
+
+    @staticmethod
+    def _get_existing_file(client: Client, context_keys: Dict[str, str]) -> Optional[File]:
+        """Find an existing File object whose context Tag matches the passed context keys"""
+        file_query = " and ".join(
+            [f'value("{key}") = "{value}"' for key, value in context_keys.items()]
+        )
+        file = File.query(client, file_query)
+        if len(file.files) == 1:
+            return file.files[0]
+        elif len(file.files) == 0:
+            return None
+        else:
+            raise SteamshipError(
+                "Multiple ChatHistory objects have been created in this workspace with these context keys."
+            )
+
+    @staticmethod
+    def get_or_create(
+        client: Client,
+        context_keys: Dict[str, str],
+        tags: List[Tag] = None,
+        initial_system_prompt: Optional[str] = None,
+    ) -> ChatHistory:
+
+        file = ChatHistory._get_existing_file(client, context_keys)
+
+        if file is None:
+            tags = tags or []
+            tags.append(Tag(kind=TagKind.DOCUMENT, name=DocTag.CHAT))
+
+            blocks = []
+            if initial_system_prompt is not None:
+                blocks.append(
+                    Block(
+                        text=initial_system_prompt,
+                        tags=[Tag(kind=TagKind.ROLE, name=RoleTag.SYSTEM)],
+                    )
+                )
+
+            file = File.create(
+                client=client,
+                blocks=blocks,
+                tags=tags,
+            )
+        return ChatHistory(file)
+
+    def append_user_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with content provided by the end-user."""
+        tags = tags or []
+        tags.append(
+            Tag(
+                kind=TagKind.CHAT, name=ChatTag.ROLE, value={TagValueKey.STRING_VALUE: RoleTag.USER}
```

Comment:
This suggestion feels like slight early-optimization, but I wonder if we want to additionally hard-code in -- right from the start -- the ability to say WHICH user or WHICH system/agent. Since eventually we might end up finding we've got a ChatHistory that's in:

* a chat room with multiple humans
* an AgentService with multiple agents